File: src/myersDiff.py
import sys

class Snake:
    __xStart = 0
    __yStart = 0
    __xEnd = 0
    __yEnd = 0
 
    def __init__(self, xStart, yStart, xEnd, yEnd):
        self.__xStart = xStart
        self.__yStart = yStart
        self.__xEnd = xEnd
        self.__yEnd = yEnd

    def getxStart(self):
        return self.__xStart
 
    def setxStart(self, xStart):
        self.__xStart = xStart
 
    def getyStart(self):
        return self.__yStart
 
    def setyStart(self, yStart):
        self.__yStart = yStart
 
    def getxEnd(self):
        return self.__xEnd

    def setxEnd(self, xEnd):
        self.__xEnd = xEnd

    def getyEnd(self):
        return self.__yEnd
 
    def setyEnd(self, yEnd):
        self.__yEnd = yEnd

def myersDiff(oldStr, newStr):
    oldContent = list(oldStr)
    newContent = list(newStr)

    lengthSum = len(oldContent) + len(newContent)
    maxStep = len(oldContent) + len(newContent)
    v = [0] * (lengthSum * 2 + 1)
    snakes = []

    for d in range(0, lengthSum + 1):
        for k in range(-d, d + 2, 2):
            print("d=", d, sep="", end=" ")
            print("k=", k, sep="", end=" ")
            down = (k == -d or (k != d and (v[k - 1 + maxStep] < v[k + 1 + maxStep])))

            if (down):
                kPrev = k + 1
            else:
                kPrev = k - 1
            xStart = v[kPrev + maxStep]
            yStart = xStart - kPrev
            if (down):
                xMid = xStart 
            else:
                xMid = xStart + 1
            yMid = xMid - k
            xEnd = xMid
            yEnd = yMid

            snake = 0
            while (xEnd < len(oldContent) and yEnd < len(newContent) and oldContent[xEnd] == newContent[yEnd]):
                xEnd += 1
                yEnd += 1
                snake += 1
            
            v[k + maxStep] = xEnd

            snakes.insert(0, Snake(xStart, yStart, xEnd, yEnd))
            print("start=(", xStart, ",", yStart, "), mid=(", xMid, ",", yMid, "), end=(", xEnd, ",", yEnd, ")", sep="")

            if (xEnd >= len(oldContent) and yEnd >= len(newContent)):
                print("Found!")
                currentSnake = snakes[0]
                print("(", currentSnake.getxEnd(), ",", currentSnake.getyEnd(), ") <- (", currentSnake.getxStart(), ",", currentSnake.getyStart(), ")", sep="")
                for i in range(1, len(snakes) - 1):
                    tmpSnake = snakes[i]
                    if (tmpSnake.getxEnd() == currentSnake.getxStart() and tmpSnake.getyEnd() == currentSnake.getyStart()):
                        currentSnake = tmpSnake
                        print("(", currentSnake.getxEnd(), ",", currentSnake.getyEnd(), ") <- (", currentSnake.getxStart(), ",", currentSnake.getyStart(), ")", sep="")
                        if ((currentSnake.getxStart() == 0) and (currentSnake.getyStart() == 0)):
                            break
                return

if __name__ == "__main__":
    try:
        oldStr= sys.argv[1]
        newStr= sys.argv[2]
        myersDiff(oldStr, newStr)
    except Exception as e:
        print(sys.argv)
        print(e)

File: src/myersDiff.js
class Snake {
    #xStart = 0
    #yStart = 0
    #xEnd = 0
    #yEnd = 0

    constructor(xStart, yStart, xEnd, yEnd) {
        this.#xStart = xStart
        this.#yStart = yStart
        this.#xEnd = xEnd
        this.#yEnd = yEnd
    }

    get getxStart(){
        return this.#xStart
    }

    set setxStart(xStart){
        this.#xStart = xStart
    }

    get getyStart(){
        return this.#yStart
    }

    set setyStart(yStart){
        this.#yStart = yStart
    }

    get getxEnd(){
        return this.#xEnd
    }

    set setxEnd(xEnd){
        this.#xEnd = xEnd
    }

    get getyEnd(){
        return this.#yEnd
    }

    set setyEnd(yEnd){
        this.#yEnd = yEnd
    }
}

function myersDiff(oldStr, newStr) {
    let oldContent = oldStr.split('')
    let newContent = newStr.split('')

    let lengthSum = oldContent.length + newContent.length
    let maxStep = oldContent.length + newContent.length
    let v = new Array(lengthSum * 2 + 1).fill(0)
    let snakes = []

    for (let d = 0; d <= lengthSum; d++) {
        for (let k = -d; k <= d; k += 2) {
            let down = (k === -d || (k !== d && (v[k - 1 + maxStep] < v[k + 1 + maxStep])))

            let kPrev = down ? k + 1 : k - 1
            let xStart = v[kPrev + maxStep]
            let yStart = xStart - kPrev
            let xMid = down ? xStart : xStart + 1
            let yMid = xMid - k
            let xEnd = xMid
            let yEnd = yMid

            let snake = 0
            while (xEnd < oldContent.length && yEnd < newContent.length && oldContent[xEnd] == newContent[yEnd]) {
                xEnd++
                yEnd++
                snake++
            }
            
            v[k + maxStep] = xEnd

            snakes.unshift(new Snake(xStart, yStart, xEnd, yEnd))
            console.log('d=' + d + ', k=' + k + ', start=(' + xStart + ',' + yStart + '), mid=(' + xMid + ',' + yMid + '), end=(' + xEnd + ',' + yEnd + ')')

            if (xEnd >= oldContent.length && yEnd >= newContent.length) {
                console.log("Found!")
                let resList = []
                let currentSnake = snakes[0]
                let outStr = '(' + currentSnake.getxEnd + ',' + currentSnake.getyEnd + ') <- (' + currentSnake.getxStart + ',' + currentSnake.getyStart + ')'
                console.log(outStr)
                resList.push(outStr)
                for (let i = 1; i < snakes.length - 1; i++) {
                    let tmpSnake = snakes[i]
                    if (tmpSnake.getxEnd === currentSnake.getxStart && tmpSnake.getyEnd === currentSnake.getyStart) {
                        currentSnake = tmpSnake
                        outStr = '(' + currentSnake.getxEnd + ',' + currentSnake.getyEnd + ') <- (' + currentSnake.getxStart + ',' + currentSnake.getyStart + ')'
                        console.log(outStr)
                        resList.push(outStr)
                        if ((currentSnake.getxStart === 0) && (currentSnake.getyStart === 0)) {
                            break
                        }
                    }
                }
                return resList
            }
        }
    }
    return []
}